jigglypuffRL/deeprl/dqn/dqn.py

```diff
@@ -0,0 +1,239 @@
+        action = Variable(torch.LongTensor(action.long()))
+        reward = Variable(torch.FloatTensor(reward))
+        done = Variable(torch.FloatTensor(done))
+
+        q_values = self.model(state)
+        if self.double_dqn:
+            next_q_values = self.target_model(next_state)
+        else:
+            next_q_values = self.model(next_state)
+
+        q_value = q_values.gather(1, action.unsqueeze(1)).squeeze(1)
+        next_q_value = next_q_values.max(1)[0]
+        expected_q_value = reward + self.gamma * next_q_value * (1 - done)
+
+        loss = F.smooth_l1_loss(q_value, expected_q_value)
+
+        self.loss_hist.append(loss)
+
+        return loss
+
+    def update_params(self):
+        loss = self.get_td_loss()
+        self.optimizer.zero_grad()
+        loss.backward()
+        self.optimizer.step()
+
+    def calculate_epsilon_by_frame(self, frame_idx):
+        return self.min_epsilon + (self.max_epsilon - self.min_epsilon) * np.exp(
+            -1.0 * frame_idx / self.epsilon_decay
+        )
+
+    def plot(self):
+        plt.figure(figsize=(20, 5))
+        plt.subplot(131)
+        plt.title(
+            "Average reward for last 10 frames: %s" % (np.mean(self.reward_hist[-10:]))
+        )
+        plt.plot(self.reward_hist)
+        plt.subplot(132)
+        plt.title("loss")
+        plt.plot(self.loss_hist)
+        plt.show()
+
+    def learn(self):
+        total_steps = self.max_epochs * self.max_iterations_per_epoch
+        state, ep_r, ep, ep_len = self.env.reset(), 0, 0, 0
+
+        if self.double_dqn:
+            self.update_target_model()
+
+        for frame_idx in range(1, total_steps + 1):
+            action = self.select_action(state, frame_idx)
+            next_state, reward, done, _ = self.env.step(action)
+
+            if self.render:
+                self.env.render()
+
+            self.replay_buffer.push((state, action, reward, next_state, done))
+
+            state = next_state
+            ep_r += reward
+            ep_len += 1
+
+            done = False if ep_len == self.max_ep_len else done
+
+            if done or (ep_len == self.max_ep_len):
+                if ep % 20 == 0:
+                    print(
+                        "Ep: {}, reward: {}, frame_idx: {}".format(ep, ep_r, frame_idx)
+                    )
+                if self.tensorboard_log:
+                    self.writer.add_scalar("episode_reward", ep_r, frame_idx)
+
+                self.reward_hist.append(ep_r)
+                state, ep_r, ep_len = self.env.reset(), 0, 0
+                ep += 1
+
+            if self.replay_buffer.get_len() > self.batch_size:
+                self.update_params()
+
+            if frame_idx % 100 == 0 and self.double_dqn:
+                self.update_target_model()
+
+        if self.plot_loss_reward_graph == True:
+            self.plot()
+
+        if self.tensorboard_log:
+            self.writer.close()
+
+        if self.render:
+            self.env.close()
```

Comment:
`self.env.close()` needs to be done regardless of whether `render` was `True` or `False`